File: lib/features/listener/domain/entities/detection_event.dart
enum DetectionEventType { helpDetected, windowReset, triggered, cooldown }

class DetectionEvent {
  const DetectionEvent({
    required this.type,
    required this.count,
    required this.timestamp,
    this.cooldownRemaining,
  });

  final DetectionEventType type;
  final int count;
  final DateTime timestamp;
  final int? cooldownRemaining;

  factory DetectionEvent.fromMap(Map<String, dynamic> map) {
    final rawType = (map['type'] as String? ?? '').toUpperCase();

    final type = switch (rawType) {
      'HELP_DETECTED' => DetectionEventType.helpDetected,
      'WINDOW_RESET' => DetectionEventType.windowReset,
      'TRIGGERED' => DetectionEventType.triggered,
      'COOLDOWN' => DetectionEventType.cooldown,
      _ => DetectionEventType.helpDetected,
    };

    final timestampMs =
        map['timestamp'] as int? ?? DateTime.now().millisecondsSinceEpoch;

    return DetectionEvent(
      type: type,
      count: map['count'] as int? ?? 0,
      timestamp: DateTime.fromMillisecondsSinceEpoch(timestampMs),
      cooldownRemaining: map['cooldownRemaining'] as int?,
    );
  }

  String summary() {
    return switch (type) {
      DetectionEventType.helpDetected =>
        'Emergency audio detected (count: $count)',
      DetectionEventType.windowReset => 'Window reset',
      DetectionEventType.triggered => 'TRIGGERED: calling emergency contact',
      DetectionEventType.cooldown =>
        'Cooldown (${cooldownRemaining ?? 0}s remaining)',
    };
  }
}

File: android/app/src/main/kotlin/com/example/sos_help_listener/DetectionEvent.kt
package com.example.sos_help_listener

enum class DetectionEventType(val wireName: String) {
    HELP_DETECTED("HELP_DETECTED"),
    WINDOW_RESET("WINDOW_RESET"),
    TRIGGERED("TRIGGERED"),
    COOLDOWN("COOLDOWN")
}

data class DetectionEvent(
    val type: DetectionEventType,
    val count: Int,
    val timestamp: Long,
    val cooldownRemaining: Int? = null
) {
    fun toMap(): Map<String, Any> {
        val payload = mutableMapOf<String, Any>(
            "type" to type.wireName,
            "count" to count,
            "timestamp" to timestamp,
        )
        cooldownRemaining?.let { payload["cooldownRemaining"] = it }
        return payload
    }
}
